Feature/617 cart item indicator

ennostuurman opened the pull request:
<!-- (REQUIRED) What is the nature of this PR? -->

## This PR is a:

- [X] New feature
- [ ] Enhancement/Optimization
- [ ] Refactor
- [ ] Bugfix
- [ ] Test for existing code
- [ ] Documentation

<!-- (REQUIRED) What does this PR change? -->
- Add visual indication to user that the cart is filled.
- Requirements see issue #617 

## Summary

When this pull request is merged, it will...
- Show an outline cart icon when cart is empty
- Show an filled cart icon when items have been added to cart

<!-- (OPTIONAL) What other information can you provide about this PR? -->

## Additional information

- did not yet add unit tests because not sure whether we want to test SVG output (filled or not).
- this functionality works fine, however I noticed when debugging propTypes.isRequired (which logs a warning on the itemsCount prop) that items_count at first render returns undefined. Looks like getCartDetails() does not return value and header component renders multiple times. Did some first tests to see if I could fix this, e.g. adding ` async componentdidMount await getCartDetails() ` . But this didn't fix it (yet).

supernova-at commented on packages/venia-concept/src/components/Header/header.js:
```diff
@@ -20,11 +22,34 @@ class Header extends Component {
+            <Icon
+                name="shopping-cart"
+                attrs={{ stroke: '#212121', fill: '#212121' }}
+            />
```
Lastly on this function, since the only thing changing is whether or not we add a `fill` attribute, we could extract that piece of logic out and simplify things a bit:

What do you think of something like:

```jsx
get cartIcon() {
    const svgAttributes = {
        stroke: 'rgb(var(--venia-text))',
    };

    if (this.props.itemsCount > 0) {
        svgAttributes.fill = 'rgb(var(--venia-text))';
    }

    return (
        <Icon
            name="shopping-cart"
            attrs={svgAttributes}
        />
    );
}
```

?

ennostuurman replied:
yeah much DRY'er ;-)

supernova-at commented on packages/venia-concept/src/components/Header/header.js:
```diff
@@ -1,4 +1,6 @@
 import React, { Component } from 'react';
+import { connect } from 'react-redux';
```
I'd like @jimbo to chime in here too but from what I've gathered so far we want to be very deliberate where we `connect` components.

In this case, an alternative approach to `connect`ing this component would be to add `itemsCount ` to the connected component `App/container.js`, and pass it down to this component via props (you'll have to pass it through `Main/main.js` too).

There's a balancing act here to be sure but it does simplify things quite nicely when done well.

ennostuurman replied:
@supernova-at thanks again for all the suggestions :+

1. I get it your point, and indeed it's a balancing act. But isn't that the reason we want to use Redux, i.e. not having to pass props down to children through components that don't require the props at all?  I find passing down props often less clear than just having a connected component in which you just 'select' a state object with mapStateToProps.

2. In a first setup I had this logic in the cartTrigger.js which is already a connected component.  I however moved the cart counter to header.js because I felt that cart counter should not be a part of cartTrigger.js which is an action based component (trigger an action)  This said I even thought about a complete separate cartCounter.js  component which might solve the header component re-render issue that I faced and which is easier to isolate / fix in a single purpose component.

Please let me know what you prefer and I'll refactor. Will refactor anyhow again because I will be adding the items counter back in again, see issue #617

ennostuurman replied:
@supernova-at see my previous comment ^^, what do you think, shall I add the counter to cartTrigger instead, so we don't introduce an additional `connect`ed component?

supernova-at replied:
Sorry for the delay, @ennostuurman - vacation called 😄 

I think absolutely a `cartCounter.js` component would be awesome.

ennostuurman replied:
@supernova-at I updated PR according to new specs, see https://github.com/magento-research/pwa-studio/issues/617#issuecomment-451650645 . 

Unfortunate a peregrine/package-lock.json made it into this PR, I always try to keep them out, this on e slipped through tho and I couldn't manage to remove it despite rebasing my branch w/ latest release/2.0 .

Could you please review? Thx!

ennostuurman replied:
@supernova-at FYI => I updated PR with latest staging, fixed header.js conflict (recent react-feather changes) .

ennostuurman commented on packages/venia-concept/src/components/Header/header.js:
```diff
@@ -20,11 +22,34 @@ class Header extends Component {
+        }),
+        // isRequired won't work because initially itemsCount is undefined
+        // seems page re-renders before cart getDetails() returns value
+        itemsCount: PropTypes.number
```
got it, good point!

supernova-at commented:
Looks like there is a conflict in `Header/header.js` as well.

supernova-at commented:
This is looking great!

Some last things:

1. We will want to remove `packages/peregrine/package-lock.json`. You can just delete it from disk, `git add .` and then commit and push that change.
1. Even when there are no items in the cart, the cart appears black / filled in for me. Is this intentional?
1. I thought I saw that once the count goes over `10` the display should show `10+`. Am I mistaken? 

Thanks again!

supernova-at commented on packages/venia-concept/src/components/Header/cartTrigger.js:
```diff
@@ -13,32 +17,47 @@ class Trigger extends Component {
+    return {
+        cart
+    };
+};
```
If you want, you can simplify this with:

```js
const mapStateToProps = ({ cart }) => ({ cart });
```

ennostuurman commented:
@supernova-at thanks for the review and suggestions. 

I made following changes:

1) refactored mapStateToProps to use destructuring as you suggested
2) added outline only layout for empty cart including test for outline / filled layout. There was some confusion around the latest designs, but Soumya confirmed outline layout, see https://github.com/magento-research/pwa-studio/issues/617#issuecomment-453596488  . 

Re you question about the counter showing 99+ . Eventually we decided to always show the actual quantity, , see https://github.com/magento-research/pwa-studio/issues/617#issuecomment-448216969 . The layout is fool proof for a 9999 counter so that should be fine.

Can you please review again? Thx!

ennostuurman commented:
hey @supernova-at I just updated this PR with latest release/2.0 branch changes. In case you have time and you're brains haven't been totally sucked up by Braintree ;-) , could you please review (see my earlier comments ;-) .? Thks!
